prompting/apparmor/message.go

```diff
@@ -0,0 +1,452 @@
+	// objects referenced by any such message.
+	Length uint16
+	// Version is the version of the communication protocol.
+	// Currently version 2 is implemented by the demo kernel based on Linux 5.7.7,
+	// but version 3 was used in the past.
+	Version uint16
+}
+
+// UnmarshalBinary unmarshals the header from binary form.
+func (msg *MsgHeader) UnmarshalBinary(data []byte) error {
+	const prefix = "cannot unmarshal apparmor message header"
+
+	// Unpack fixed-size elements.
+	order := binary.LittleEndian
+	buf := bytes.NewBuffer(data)
+	if err := binary.Read(buf, order, msg); err != nil {
+		return xerrors.Errorf("%s: %s", prefix, err)
+	}
+
+	if msg.Version != 2 {
+		return xerrors.Errorf("%s: unsupported version: %d", prefix, msg.Version)
+	}
+	if int(msg.Length) != len(data) {
+		return xerrors.Errorf("%s: length mismatch %d != %d",
+			prefix, msg.Length, len(data))
+	}
+
+	return nil
+}
+
+// RequestBuffer returns a new buffer for communication with the kernel.
+// The buffer contains encoded information about its size and protocol version.
+func RequestBuffer() []byte {
+	buf := make([]byte, 0xFFFF)
+	header := MsgHeader{Version: 2, Length: uint16(len(buf))}
+	binary.Write(&overwrite{Buffer: buf}, binary.LittleEndian, &header)
+	return buf
+}
+
+// msgNotificationFilter describes the configuration of kernel-side message filtering.
+//
+// This structure corresponds to the kernel type struct apparmor_notif_filter
+// described below. This type is only used for message marshaling and
+// unmarshaling. Application code should use MsgNotificationFilter instead.
+//
+//	struct apparmor_notif_filter {
+//	  struct apparmor_notif_common base;
+//	  __u32 modeset;      /* which notification mode */
+//	  __u32 ns;           /* offset into data, relative to start of the structure */
+//	  __u32 filter;       /* offset into data, relative to start of the structure */
+//	  __u8 data[];
+//	} __attribute__((packed));
+type msgNotificationFilter struct {
+	MsgHeader
+	ModeSet uint32
+	NS      uint32
+	Filter  uint32
+}
+
+// MsgNotificationFilter describes the configuration of kernel-side message filtering.
+//
+// This structure can be marshaled and unmarshaled to binary form and
+// transmitted to the kernel using NotifyIoctl along with IoctlGetFilter and
+// IoctlSetFilter.
+type MsgNotificationFilter struct {
+	MsgHeader
+	// ModeSet is a bitmask. Specifying ModeSetUser allows to receive notification
+	// messages in userspace.
+	ModeSet ModeSet
+	// XXX: This is currently unused by the kernel and the value is ignored.
+	NameSpace string
+	// XXX: This is currently unused by the kernel and the format is unknown.
+	Filter string
+}
+
+// UnmarshalBinary unmarshals the message from binary form.
+func (msg *MsgNotificationFilter) UnmarshalBinary(data []byte) error {
+	const prefix = "cannot unmarshal apparmor notification filter message"
+
+	// Unpack the base structure.
+	if err := msg.MsgHeader.UnmarshalBinary(data); err != nil {
+		return err
```

Comment:
I think the idea was that the error is unrelated to the fact that it's being called from within `MsgNotificationFilter.UnmarshalBinary()` instead of `MsgHeader.UnmarshalBinary()`, so that error should be returned directly instead of with a prefix. Alternatively, including the prefix followed by the `MsgHeader` error message (that is, `fmt.Errorf("%s: %v", prefix, err)` would give an indication of where `MsgHeader.UnmarshalBinary()` was being called from, but it might just distract from the error itself. I'm unsure which I prefer...

Reply:
I've made a change to add the prefix, I think it adds some clarity and consistency.